daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/Value.scala

```diff
@@ -358,7 +359,7 @@ object Value extends CidContainer1WithDefaultCidResolver[Value] {
     * to be able to use AbsoluteContractId elsewhere, so that we can
     * automatically upcast to ContractId by subtyping.
     */
-  sealed abstract class ContractId
+  sealed abstract class ContractId extends Equals
 
   sealed abstract class AbsoluteContractId extends ContractId with Product with Serializable {
```

Comment:
With `case class` on all leaves this can percolate all the way up.

```suggestion
  sealed abstract class ContractId extends Product with Serializable

  sealed abstract class AbsoluteContractId extends ContractId {
```